File: javascript/packages/core/src/models/customer.js
/** Customer record. */

import { withReference } from "../concerns/hasReference.js";

/**
 * Owner of one or more devices.
 *
 * `reference()` comes from the mixin installed at the foot of this file, not
 * from the class body.
 */
export class Customer {
    /**
     * @param {number} id
     * @param {string} name
     * @param {string} email
     * @param {string} [phone]
     */
    constructor(id, name, email, phone) {
        this.id = id;
        this.name = name;
        this.email = email;
        this.phone = phone;
    }

    /**
     * Display name used in tables and the report header.
     *
     * @returns {string}
     */
    displayName() {
        return `${this.name} <${this.email}>`;
    }

    /** @returns {boolean} */
    isReachable() {
        return this.phone !== undefined || this.email !== "";
    }

    /**
     * @param {number} id
     * @param {string} name
     * @param {string} email
     * @param {string} [phone]
     * @returns {Customer}
     */
    static seed(id, name, email, phone) {
        const customer = new Customer(id, name, email, phone);
        customer.referenceNumber = id;

        return customer;
    }
}

withReference(Customer, "CU");

File: rust/crates/atelier-core/src/models/customer.rs
//! Customer record.

use crate::concerns::has_reference::HasReference;
use serde::{Deserialize, Serialize};

/// Owner of one or more devices.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Customer {
    pub id: u32,
    pub name: String,
    pub email: String,
    pub phone: Option<String>,
    pub reference_number: u32,
}

impl Customer {
    pub fn new(id: u32, name: impl Into<String>, email: impl Into<String>) -> Self {
        Self {
            id,
            name: name.into(),
            email: email.into(),
            phone: None,
            reference_number: id,
        }
    }

    pub fn with_phone(mut self, phone: impl Into<String>) -> Self {
        self.phone = Some(phone.into());
        self
    }

    /// Display name used in tables and the report header.
    pub fn display_name(&self) -> String {
        format!("{} <{}>", self.name, self.email)
    }

    pub fn is_reachable(&self) -> bool {
        self.phone.is_some() || !self.email.is_empty()
    }
}

impl HasReference for Customer {
    const PREFIX: &'static str = "CU";

    fn reference_number(&self) -> u32 {
        self.reference_number
    }
}
